### PR TITLE
[Feature] Implement global transaction id & support a simple interface for query by gtid for optimize table

### PR DESCRIPTION
**Pull request by meegoo**
## Why I'm doing:
This is the basic capability for future implementations of multi-statement transactions, time travel queries, and other features. Currently, it also provides a simpler syntax to support online optimize table for data rewrites of specified versions cross multi partition.

## What I'm doing:
Implement the data part of the global transaction id (GTID), and also implement a simple SQL query interface based on GTID or timestamp to support data rewrite operations in online optimize table. 
The syntax is similar to `select * from table before '2020-01-01 00:00:00'`. Currently, the capability for complete time travel queries is not supported, so this interface is not exposed to users. The syntax may be adjusted in the future.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/tablet_updates.cpp [hunk 1]
```diff
@@ -1959,9 +1975,11 @@ Status TabletUpdates::_commit_compaction(std::unique_ptr<CompactionInfo>* pinfo,
     std::unique_ptr<EditVersionInfo> edit_version_info = std::make_unique<EditVersionInfo>();
     edit_version_info->version = EditVersion(edit_version_pb->major_number(), edit_version_pb->minor_number());
     edit_version_info->creation_time = creation_time;
+    edit_version_info->gtid = rowset->rowset_meta()->gtid();
```
Comment (wyb) [inline be/src/storage/tablet_updates.cpp:1978]: the input rowsets in primary table compaction may be discontinuous, so the output rowset `gtid` may be inaccurate.
the `gtid` in edit_version_info should be unchanged after compaction.

File: fe/fe-core/src/main/java/com/starrocks/transaction/GtidGenerator.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+package com.starrocks.transaction;
+
+import com.google.gson.annotations.SerializedName;
+import com.starrocks.persist.gson.GsonPostProcessable;
+
+import java.io.IOException;
+
+public class GtidGenerator implements GsonPostProcessable {
+    // |-- 1bit --|-- 42bit --|-- 8bit --|-- 13bit --|
+    // |    0     | timestamp |  cluster |  sequence |
```
[the remark this answers is not in the document]
Comment (meegoo) [reply fe/fe-core/src/main/java/com/starrocks/transaction/GtidGenerator.java:27]: first bit mark gtid is overflow or not
gtid should be order by timestamp, even if we need sync across clusters.

File: be/src/storage/tablet_updates.cpp [hunk 3]
```diff
@@ -1798,6 +1812,8 @@ Status TabletUpdates::_do_compaction(std::unique_ptr<CompactionInfo>* pinfo) {
             CompactionUtils::get_segment_max_rows(config::max_segment_file_size, input_row_num, input_rowsets_size);
     context.writer_type =
             (algorithm == VERTICAL_COMPACTION ? RowsetWriterType::kVertical : RowsetWriterType::kHorizontal);
+    // use max gtid as the gtid of new rowset
+    context.gtid = max_gtid;
```
Comment (wyb) [inline be/src/storage/tablet_updates.cpp:1816]: not used

File: fe/fe-core/src/main/java/com/starrocks/transaction/GtidGenerator.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    public GtidGenerator() {
+        nextGtid();
+    }
+
+    public synchronized long nextGtid() {
+        long timestamp = timeGen();
+
+        if (timestamp < lastTimestamp) {
+            // 使用上一个 GTID 的时间戳
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/transaction/GtidGenerator.java:47]: eng?

**Review by wyb — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**